File: src/209.长度最小的子数组.cpp
/*
 * @lc app=leetcode.cn id=209 lang=cpp
 *
 * [209] 长度最小的子数组
 */
#include <climits>
#include <iostream>
#include <vector>
using namespace std;

// @lc code=start
class Solution {
public:
    int minSubArrayLen(int target, vector<int>& nums)
    {
        int res = INT_MAX;
        int sum = 0;
        int start = 0;
        for (int fast = 0; fast < nums.size(); fast++) {
            sum += nums[fast];
            while (sum >= target) {
                if (res > fast - start + 1) {
                    res = fast - start + 1;
                }
                sum -= nums[start];
                start++;
            }
        }
        return res < INT_MAX ? res : 0;
    }
};
// @lc code=end
int main()
{
    Solution s;
    int target = 15;
    vector<int> nums = { 5, 1, 3, 5, 10, 7, 4, 9, 2, 8 };
    int res = s.minSubArrayLen(target, nums);
    cout << res << endl;
}

File: 数组/209.长度最小的子数组.py
from typing import List


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        """关键在于双指针对应的滑动窗口

        Args:
            target (int): _description_
            nums (List[int]): _description_

        Returns:
            int: _description_
        """
        slow = 0
        fast = 0
        cur = 0
        res = float("inf")
        while fast < len(nums):
            cur += nums[fast]
            while cur >= target:
                res = min(res, fast - slow + 1)
                cur -= nums[slow]
                slow += 1
            fast += 1
        return res if res != float("inf") else 0


if __name__ == "__main__":
    nums = [2, 3, 1, 2, 4, 3]
    print(Solution().minSubArrayLen(7, nums))
